src/dm/DispatchManagerActions.cc

```diff
@@ -2249,7 +2303,750 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+
+            vm->get_template_attribute("MEMORY", omemory);
+            vm->get_template_attribute("CPU", ocpu);
+            vm->get_template_attribute("VCPU", ovcpu);
+
+            if (ocpu == cpu && omemory == memory && ovcpu == vcpu)
+            {
+                rc = 0;
+                error_str = "Nothing to resize";
+                break;
+            }
+
+            rc = test_set_capacity(vm.get(), cpu, memory, vcpu, enforce, error_str);
+
+            if (rc == 0)
+            {
+                vm->set_state(VirtualMachine::HOTPLUG_RESIZE);
+
+                vm->store_resize(ocpu, omemory, ovcpu);
+
+                vm->set_resched(false);
+
+                vmm->trigger_resize(vid);
+            }
+            break;
+        }
+
+        case VirtualMachine::STOPPED:
+        case VirtualMachine::DONE:
+        case VirtualMachine::SUSPENDED:
+            rc = -1;
+            error_str = "Cannot resize a VM in state " + vm->state_str();
+            break;
+    }
+
+    if (rc == 0)
+    {
+        if (vm->hasHistory())
+        {
+            close_cp_history(vmpool, vm.get(), VMActions::RESIZE_ACTION, ra);
+        }
+
+        vmpool->update(vm.get());
+    }
+
+    return rc;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int DispatchManager::attach_pci(int vid, VectorAttribute * pci,
+                                const RequestAttributes& ra, string& error_str)
+{
+    auto vm = vmpool->get(vid);
+
+    if ( vm == nullptr )
+    {
+        error_str = "Virtual machine does not exist";
+        return -1;
+    }
+
+    if ((vm->get_state() != VirtualMachine::POWEROFF &&
+         vm->get_state() != VirtualMachine::UNDEPLOYED) ||
+        vm->get_lcm_state() != VirtualMachine::LCM_INIT)
+    {
+        error_str = "VM in wrong state, it has to be in poweroff";
+        return -1;
+    }
+
+    int hid;
+
+    if (vm->hasHistory())
+    {
+        hid = vm->get_hid();
+    }
+    else
+    {
+        error_str = "Could not find host information";
+        return -1;
+    }
+
+    unique_ptr<Host> host;
+
+    if (vm->get_state() ==  VirtualMachine::POWEROFF)
+    {
+        HostShareCapacity sr;
+
+        sr.vmid = vid;
+        sr.pci.push_back(pci);
+
+        host = hpool->get(hid);
+
+        if ( host == nullptr )
+        {
+            error_str = "Could not find host information";
+            return -1;
+        }
+
+        if (!host->add_pci(sr))
+        {
+            error_str = "Cannot assign PCI device in host. Check address and free devices";
+            return -1;
+        }
+    }
+
+    if ( vm->attach_pci(pci, error_str) == -1 )
+    {
+        return -1;
+    }
+
+    if (host)
+    {
+        hpool->update(host.get());
+    }
+
+    close_cp_history(vmpool, vm.get(), VMActions::PCI_ATTACH_ACTION, ra);
+
+    vm->log("DiM", Log::INFO, "PCI device successfully attached.");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential log injection detected. Ensure all untrusted input is properly sanitized before logging. Use parameterized logging or validate input against an allow list to prevent log injection vulnerabilities. Consider using a dedicated logging library&#39;s built-in sanitization features when available. Learn more - https://cwe.mitre.org/data/definitions/117.html